File: cpp/trex_selector_methods/trex/demo_trex_04_mc_sim_lloop_strategies/demo_trex_04_mc_sim_lloop_strategies.cpp
// ==============================================================================
// demo_trex_04_lloop_strategies.cpp
// ==============================================================================
/**
 * @file demo_trex_04_lloop_strategies.cpp
 *
 * @brief Monte Carlo simulation comparing the six L-loop strategies of the
 *        T-Rex Selector, with TLARS held fixed as the base solver.
 *
 * @details Mirrors R/trex_selector_methods/trex/demo_trex_04_lloop_strategies.R
 *
 *  Runs in two modes controlled by the `block_support` parameter:
 *    block_support = true  — contiguous block support {0, 1, ..., s-1}.
 *    block_support = false — scattered support redrawn from uniform per trial.
 *
 *  Holds the solver fixed (TLARS) and sweeps over the six L-loop strategies:
 *
 *  Demo content:
 *    STANDARD           — Fresh i.i.d. dummy matrix at each L-loop iteration.
 *    HCONCAT            — Horizontally expand dummy columns.
 *    PERMUTATION        — Re-use the base dummy matrix via column permutations.
 *    PERMUTATION_DIRECT — Seed-based permutations; no base matrix in memory.
 *    DIRECT             — Seed-based i.i.d. draws; no base matrix in memory.
 *    SKIPL              — Skip the L-loop entirely and uses L = max_dummy_multiplier.
 *
 *    High-dimensional  (n = 300, p = 1000, s = 10).
 *    SNR sweep: {0.1, 0.2, ..., 2.0, 5.0}  (21 values).
 *    Two support scenarios:
 *      random support  — redrawn per trial;
 *      block support   — contiguous block {0, 1, ..., s-1}.
 *    Reports averaged FDR / TPR / Avg L / Avg T per L-strategy x SNR.
 */
// ==============================================================================

// std includes
#include <algorithm>
#include <iomanip>
#include <iostream>
#include <map>
#include <numeric>
#include <random>
#include <sstream>
#include <string>
#include <vector>

// Eigen includes
#include <Eigen/Dense>

// OpenMP compatibility layer
#include <utils/openmp/utils_openmp.hpp>

// T-Rex Selector includes
#include <trex_selector_methods/trex_core/trex.hpp>
#include <utils/datageneration/utils_datagen.hpp>
#include <utils/eval_metrics/utils_eval_cdiagnostics.hpp>
#include <utils/eval_metrics/utils_eval_rates.hpp>

// Demo utilities
#include "trex_sim_utils.hpp"
using namespace trex_sim;


// ==============================================================================
// Namespace aliases
// ==============================================================================

namespace cdianostics = trex::utils::eval::cdiagnostics;
namespace datagen = trex::utils::datageneration::datagen;
namespace dummygen = trex::utils::datageneration::dummygen;

// T-Rex types
using trex::trex_selector_methods::trex_core::TRexControlParameter;
using trex::trex_selector_methods::trex_core::LLoopStrategy;
using trex::trex_selector_methods::utils::solver_dispatch::SolverTypeForTRex;


// ==============================================================================
// L-loop strategy descriptor
// ==============================================================================

/**
 * @brief Descriptor for one L-loop strategy variant.
 */
struct LLoopStrategyInfo {
    /** @brief Row label used in output table and CSV. */
    std::string name;
    /** @brief Enum value passed to TRexControlParameter. */
    LLoopStrategy strategy;
    /**
     * @brief Upper bound for the adaptive L-loop (non-SKIPL) or the fixed L
     *        value for SKIPL (expressed as a multiple of p, so total dummies
     *        = max_dummy_multiplier * p).
     */
    std::size_t max_dummy_multiplier = 10;
};


/** @brief Returns the canonical list of L-loop strategy variants to compare.
 *
 *  Adaptive strategies (STANDARD … DIRECT) all use max_dummy_multiplier = 10 (default).
 *  SKIPL is compared at four explicit L levels: 5p, 10p, 20p, and 50p.
 */
static std::vector<LLoopStrategyInfo> make_lloop_strategies() {
    return {
        //{"STANDARD",           LLoopStrategy::STANDARD},
        //{"HCONCAT",            LLoopStrategy::HCONCAT},
        //{"PERMUTATION",        LLoopStrategy::PERMUTATION},
        //{"PERMUTATION_DIRECT", LLoopStrategy::PERMUTATION_DIRECT},
        //{"DIRECT",             LLoopStrategy::DIRECT},
        {"SKIPL_5p",           LLoopStrategy::SKIPL, 5},
        {"SKIPL_10p",          LLoopStrategy::SKIPL, 10},
        {"SKIPL_20p",          LLoopStrategy::SKIPL, 20}
    };
}


// ==============================================================================
// Monte Carlo Simulation — L-loop strategy comparison (TLARS fixed)
// ==============================================================================

void demo_TRexSelector_lloop_strategies(std::size_t num_MC, bool rnd_coef,
                                        bool block_support) {

    std::cout.setf(std::ios::unitbuf);

    cdianostics::print_section_header(
        "Demo: T-Rex Selector MC \u2014 L-Strategy Comparison  |  TLARS");

    // -----------------------------------------------------------------------
    // Scenario parameters (always high-dimensional, mirrors R file)
    // -----------------------------------------------------------------------
    const Eigen::Index n = 300;
    const Eigen::Index p = 1000;
    const std::size_t  s = 10;
    const double       tFDR = 0.1;

    std::cout << "High-dimensional (p > n)\n";
    std::cout << "n = " << n << ",  p = " << p
              << ",  s = " << s << ",  num_MC = " << num_MC
              << "  [" << (block_support ? "block" : "random") << " support]\n\n";

    // SNR values: 0.1, 0.2, ..., 2.0, 5.0  (21 values)
    std::vector<double> snr_values(20);
    std::iota(snr_values.begin(), snr_values.end(), 1);
    for (auto& x : snr_values) x *= 0.1;
    snr_values.push_back(5.0);

    // -----------------------------------------------------------------------
    // L-loop strategies to sweep
    // -----------------------------------------------------------------------
    const std::vector<LLoopStrategyInfo> strategies = make_lloop_strategies();

    // -----------------------------------------------------------------------
    // Result containers: strategy x SNR
    // -----------------------------------------------------------------------
    std::map<std::string, Eigen::VectorXd> fdr_results_map;
    std::map<std::string, Eigen::VectorXd> tpr_results_map;
    std::map<std::string, Eigen::VectorXd> average_L_results_map;
    std::map<std::string, Eigen::VectorXd> average_T_results_map;

    std::vector<std::string> strategy_names;
    strategy_names.reserve(strategies.size());
    for (const auto& strat : strategies) {
        strategy_names.push_back(strat.name);
        fdr_results_map[strat.name]       = Eigen::VectorXd(snr_values.size());
        tpr_results_map[strat.name]       = Eigen::VectorXd(snr_values.size());
        average_L_results_map[strat.name] = Eigen::VectorXd(snr_values.size());
        average_T_results_map[strat.name] = Eigen::VectorXd(snr_values.size());
    }

    // -----------------------------------------------------------------------
    // Strategy loop
    // -----------------------------------------------------------------------
    for (const auto& strat : strategies) {

        cdianostics::print_section_header("L-strategy: " + strat.name);

        TRexControlParameter ctrl;
        ctrl.solver_type          = SolverTypeForTRex::TLARS;
        ctrl.K                    = 20;
        ctrl.max_dummy_multiplier = strat.max_dummy_multiplier;
        ctrl.use_max_T_stop       = true;
        ctrl.dummy_distribution   = dummygen::Distribution::Normal();
        ctrl.lloop_strategy       = strat.strategy;

        // Simulation loop: sweep over SNR values
        for (std::size_t snr_idx = 0; snr_idx < snr_values.size(); ++snr_idx) {

            const double   snr       = snr_values[snr_idx];
            const unsigned base_seed = 24u + static_cast<unsigned>(snr_idx) * 1000u;

            std::cout << "SNR = " << std::fixed << std::setprecision(2) << snr << "\n";

            const auto make_data = [=](unsigned seed) -> trex_sim::TrexDGPData {
                std::vector<std::size_t> pool;
                if (block_support) {
                    pool.resize(s);
                    std::iota(pool.begin(), pool.end(), std::size_t{0});
                } else {
                    std::mt19937 rng_sup(seed + 500000u);
                    pool.resize(static_cast<std::size_t>(p));
                    std::iota(pool.begin(), pool.end(), std::size_t{0});
                    std::shuffle(pool.begin(), pool.end(), rng_sup);
                    pool.resize(s);
                    std::sort(pool.begin(), pool.end());
                }

                std::vector<double> coefs;
                coefs.reserve(s);
                if (rnd_coef) {
                    std::mt19937 rng_coef(seed + 600000u);
                    std::normal_distribution<double> nd(0.0, 1.0);
                    for (std::size_t i = 0; i < s; ++i)
                        coefs.push_back(nd(rng_coef));
                } else {
                    for (std::size_t i = 0; i < s; ++i)
                        coefs.push_back(1.0);
                }

                datagen::SyntheticData data(
                    n, p,
                    pool, coefs, snr,
                    static_cast<int>(seed),
                    datagen::predictor_policy::Normal(),
                    datagen::noisegen::noise_policy::Normal());

                return {data.getX(), data.getY(), pool};
            };

            std::ostringstream lbl;
            lbl << "SNR=" << std::fixed << std::setprecision(2) << snr
                << " [" << strat.name << "]";
            const auto result = trex_sim::run_mc_trials_trex(
                num_MC, lbl.str(), make_data, tFDR, ctrl,
                base_seed);

            const auto ei = static_cast<Eigen::Index>(snr_idx);
            fdr_results_map[strat.name](ei)        = result.avg_fdr;
            tpr_results_map[strat.name](ei)        = result.avg_tpr;
            average_L_results_map[strat.name](ei)  = result.avg_L;
            average_T_results_map[strat.name](ei)  = result.avg_T;
        }

        std::cout << "\n";
    }

    // -----------------------------------------------------------------------
    // Print and save results
    // -----------------------------------------------------------------------
    const std::string stem = "demo_trex_04_lloop_strategies_results_n"
        + std::to_string(static_cast<std::size_t>(n))
        + "_p" + std::to_string(static_cast<std::size_t>(p))
        + (block_support ? "_block_support" : "_random_support");

    save_and_print_mc_results(
        num_MC,
        stem,
        snr_values,
        strategy_names,
        fdr_results_map,
        tpr_results_map,
        average_L_results_map,
        average_T_results_map);

    std::cout << "\n\n";
}


// ==============================================================================
// Main
// ==============================================================================

int main() {

    std::cout.setf(std::ios::unitbuf);
    omp_set_num_threads(6);
    std::cout << "Running with " << omp_get_max_threads() << " threads\n\n";

    if (true)
        demo_TRexSelector_lloop_strategies(/*num_MC=*/10,
                                           /*rnd_coef=*/false,
                                           /*block_support=*/false);
    if (false)
        demo_TRexSelector_lloop_strategies(/*num_MC=*/10,
                                           /*rnd_coef=*/false,
                                           /*block_support=*/true);
    return 0;
}

File: Python/trex_selector_methods/trex/demo_trex_04_mc_sim_lloop_strategies/demo_trex_04_mc_sim_lloop_strategies.py
# ==============================================================================
# demo_trex_04_mc_sim_lloop_strategies.py
# ==============================================================================
#
# Classical T-Rex selector demo — L-loop strategy comparison, repeated across
# three base solvers (TLARS / TOMP / TAFS).
# Mirrors cpp/trex_selector_methods/trex/demo_trex_04_mc_sim_lloop_strategies.cpp
# (which extends R/trex_selector_methods/trex/demo_trex_04_mc_sim_lloop_strategies.R;
# the R version fixes the base solver at TLARS).
#
# L-loop strategies (8 rows):
#
#    STANDARD             — Stored; fresh i.i.d. dummy matrix at each L-loop
#                           iteration (conservative default).
#    HCONCAT              — Stored; horizontally expand dummy columns from the
#                           same base draw (faster, prefix-stable).
#    PERMUTATION          — Stored base dummy matrix re-used via deterministic
#                           row permutations per experiment.
#    PERMUTATION_ONDEMAND — Seed-derived base + row permutations; nothing
#                           stored. Bit-identical to PERMUTATION per seed.
#    ONDEMAND             — Seed-derived i.i.d. draws; nothing stored.
#    SKIPL_{5p,10p,20p}   — Skip the L-loop entirely; fixed
#                           L = max_dummy_multiplier in {5, 10, 20} x p.
#
#  Scenario:
#  High-dimensional (n = 300, p = 1000, s = 10).
#  SNR sweep: {0.1, 0.2, ..., 2.0, 5.0}  (21 values).
#  Base solvers (outer sweep, one result file pair per solver):
#    TLARS, TOMP, TAFS (rho_afs = 0.3).
#  Two support scenarios:
#    random support — redrawn per trial;
#    block support  — contiguous block {0, 1, ..., s-1}  (0-based).
#  Reports averaged FDR / TPR / Avg L / Avg T per L-strategy x SNR.
#
# Python downscale: num_MC = 25 (C++ runs 200) — the 3-solver x 8-strategy x
# 21-SNR grid is large; increase _NUM_MC for publication-grade curves.
#
# ==============================================================================

import sys
import os

_THIS_DIR = os.path.dirname(os.path.abspath(__file__))
_PARENT_DIR = os.path.dirname(_THIS_DIR)
for _p in (_THIS_DIR, _PARENT_DIR):
    if _p not in sys.path:
        sys.path.insert(0, _p)

import numpy as np

from trex_sim_common import run_mc_trex, save_mc_results

# ==============================================================================
# Global parameters
# ==============================================================================

_OUT_DIR     = os.path.join(_THIS_DIR, "simulation_results", "data")
_NUM_WORKERS = 6
_NUM_MC      = 25

# ==============================================================================
# L-loop strategies to compare (mirrors make_lloop_strategies() in C++)
# ==============================================================================
#
# name           : row label in output tables and CSV.
# strategy       : string passed to lloop_strategy in _make_trex_ctrl_from_dict.
# max_dummy_mult : upper bound for the adaptive L-loop (non-SKIPL) or the fixed
#                  L value for SKIPL (as a multiple of p).

def _strat(name, strategy, max_dummy_mult=10):
    return dict(name=name, strategy=strategy, max_dummy_mult=max_dummy_mult)

L_STRATEGIES = [
    _strat("STANDARD",             "STANDARD"),
    _strat("HCONCAT",              "HCONCAT"),
    _strat("PERMUTATION",          "PERMUTATION"),
    _strat("PERMUTATION_ONDEMAND", "PERMUTATION_ONDEMAND"),
    _strat("ONDEMAND",             "ONDEMAND"),
    _strat("SKIPL_5p",  "SKIPL",  5),
    _strat("SKIPL_10p", "SKIPL", 10),
    _strat("SKIPL_20p", "SKIPL", 20),
]

# Base solvers for the outer sweep (rho_afs only meaningful for TAFS).
# Deviations between the solvers will be detectable in the results.
_SOLVERS = [
    dict(solver_name="TLARS", name="TLARS", lambda2=0.1, rho_afs=0.3, ncgmp_variant=0),
    dict(solver_name="TOMP",  name="TOMP",  lambda2=0.1, rho_afs=0.3, ncgmp_variant=0),
    dict(solver_name="TAFS",  name="TAFS",  lambda2=0.1, rho_afs=0.3, ncgmp_variant=0),
]


# ==============================================================================
# MC simulation — L-loop strategy comparison
# ==============================================================================

def trx_04_lloop_strategies(num_MC=_NUM_MC, num_workers=_NUM_WORKERS,
                            rnd_coef=False, block_support=False):

    n = 300
    p = 1000
    s = 10
    tFDR       = 0.1
    snr_values = list(np.round(np.arange(0.1, 2.1, 0.1), 1)) + [5.0]  # 21 values

    support_label = "block" if block_support else "random"
    print(f"\n{'=' * 70}")
    print("MC Simulation — L-Strategy Comparison  |  TLARS / TOMP / TAFS")
    print("  High-dimensional (p > n)")
    print(f"  n = {n},  p = {p},  s = {s},  num_MC = {num_MC}  [{support_label} support]\n")

    # Block support: contiguous 0-based {0, 1, ..., s-1}  (fixed across all trials)
    block_supp = np.arange(s, dtype=np.intp) if block_support else None

    n_strat = len(L_STRATEGIES)
    n_snr   = len(snr_values)
    strat_names = [st["name"] for st in L_STRATEGIES]

    # Solver sweep — one result file pair (txt + csv) per base solver
    for solver_desc in _SOLVERS:

        # Result containers: strategy x SNR (reset per solver)
        fdr_mat   = np.zeros((n_strat, n_snr))
        tpr_mat   = np.zeros((n_strat, n_snr))
        avg_L_mat = np.zeros((n_strat, n_snr))
        avg_T_mat = np.zeros((n_strat, n_snr))

        # Strategy sweep: iterate over L-loop strategies and SNR values
        for si, strat in enumerate(L_STRATEGIES):
            print(f"{'=' * 50}")
            print(f"Solver: {solver_desc['name']}  |  L-strategy: {strat['name']}")
            print(f"{'=' * 50}\n")

            # Per-strategy ctrl — max_dummy_multiplier and lloop_strategy vary
            base_ctrl = dict(
                K                        = 20,
                max_dummy_multiplier     = strat["max_dummy_mult"],
                use_max_T_stop           = True,
                lloop_strategy           = strat["strategy"],
                parallel_rnd_experiments = False,   # mirrors R use_openmp=FALSE
            )

            for snr_idx, snr in enumerate(snr_values):
                base_seed = 24 + snr_idx * 1000
                lbl = f"SNR={snr:.2f} [{strat['name']}]"
                result = run_mc_trex(
                    n=n, p=p, s=s, snr=snr,
                    solver_desc=solver_desc,
                    base_ctrl_dict=base_ctrl,
                    tFDR=tFDR,
                    base_seed=base_seed,
                    num_MC=num_MC,
                    max_workers=num_workers,
                    fixed_support=block_supp,   # None → random per trial
                    amplitude=1.0,
                    rnd_coef=rnd_coef,
                    track_L_T=True,
                    label=lbl,
                )
                fdr_mat[si, snr_idx]   = result["mean_FDR"]
                tpr_mat[si, snr_idx]   = result["mean_TPR"]
                avg_L_mat[si, snr_idx] = result["mean_L"]
                avg_T_mat[si, snr_idx] = result["mean_T"]
            print()

        stem = (f"demo_trex_04_lloop_strategies_results_n{n}_p{p}"
                f"_{solver_desc['name'].lower()}_{support_label}_support")
        save_mc_results(
            _OUT_DIR, stem, num_MC,
            strat_names, snr_values,
            fdr_mat, tpr_mat, avg_L_mat, avg_T_mat,
        )
        print()


# ==============================================================================
# Main
# ==============================================================================

if __name__ == "__main__":
    trx_04_lloop_strategies(block_support=False)
    # trx_04_lloop_strategies(block_support=True)
    print("\ndemo_trex_04_mc_sim_lloop_strategies.py complete.")
